src/optimistic_transaction_db.rs

```diff
@@ -0,0 +1,259 @@
+            {
+                cfs.push(ColumnFamilyDescriptor {
+                    name: String::from("default"),
+                    options: Options::default(),
+                });
+            }
+            // We need to store our CStrings in an intermediate vector
+            // so that their pointers remain valid.
+            let c_cfs: Vec<CString> = cfs
+                .iter()
+                .map(|cf| CString::new(cf.name.as_bytes()).unwrap())
+                .collect();
+
+            let mut cf_names: Vec<_> = c_cfs.iter().map(|cf| cf.as_ptr()).collect();
+
+            // These handles will be populated by DB.
+            let mut cf_handles: Vec<_> = cfs.iter().map(|_| ptr::null_mut()).collect();
+
+            let mut cf_opts: Vec<_> = cfs.iter().map(|cf| cf.options.inner as *const _).collect();
+
+            unsafe {
+                db = ffi_try!(ffi::rocksdb_optimistictransactiondb_open_column_families(
+                    opts.inner,
+                    cpath.as_ptr(),
+                    cfs.len() as c_int,
+                    cf_names.as_mut_ptr(),
+                    cf_opts.as_mut_ptr(),
+                    cf_handles.as_mut_ptr(),
+                ));
+            }
+
+            for handle in &cf_handles {
+                if handle.is_null() {
+                    return Err(Error::new(
+                        "Received null column family \
+                         handle from DB."
+                            .to_owned(),
+                    ));
+                }
+            }
+
+            for (n, h) in cfs.iter().zip(cf_handles) {
+                cf_map
+                    .write()
+                    .map_err(|e| Error::new(e.to_string()))?
+                    .insert(n.name.clone(), h);
+            }
+        }
+
+        if db.is_null() {
+            return Err(Error::new("Could not initialize database.".to_owned()));
+        };
+
+        let path_buf = path.to_path_buf();
+
+        let base_db = unsafe {
+            let inner = ffi::rocksdb_optimistictransactiondb_get_base_db(db);
+            DB {
+                inner,
+                cfs: cf_map,
+                path: path_buf.clone(),
+                is_base_db: true,
+            }
+        };
+
+        Ok(OptimisticTransactionDB {
+            inner: db,
+            path: path_buf,
+            base_db,
+        })
+    }
+
+    /// Begins a new optimistic transaction.
+    pub fn transaction(
+        &self,
+        write_options: &WriteOptions,
+        optimistic_tx_options: &OptimisticTransactionOptions,
+    ) -> Transaction {
+        unsafe {
+            let inner = ffi::rocksdb_optimistictransaction_begin(
+                self.inner,
+                write_options.inner,
+                optimistic_tx_options.inner,
+                ptr::null_mut(),
+            );
+            let snapshot = if optimistic_tx_options.set_snapshot {
+                Some(ffi::rocksdb_transaction_get_snapshot(inner))
```

Comment:
Will RocksDB manage the lifetime of this snapshot, or does something need to explicitly `close` or `delete` it?

Reply:
https://github.com/facebook/rocksdb/blob/c2affccc1839297b097e847cb539103e96a66e5d/utilities/transactions/transaction_base.cc#L35

Reply:
https://github.com/facebook/rocksdb/issues/4999

Reply:
no api delete the rocksdb_snapshot_t.